### PR TITLE
1299 do not show datanames error if there are transformers passed to modules

### PR DESCRIPTION
**Pull request by m7pr**
Fixes the top part of https://github.com/insightsengineering/teal/issues/1299

The current behavior raised a warning that datanames are not sufficient based on comparison of `data@datanames` and `modules$datanames`. This now have a new condition - warning is only displayed when the above is satisfied & there are no `transformers`.

image

# Code for testing

<details><summary>Tested with</summary>

```r
pkgload::load_all('teal.data')
pkgload::load_all('teal.slice')
pkgload::load_all('teal')

my_transformers <- list(
  teal_transform_module(
    label = "Custom transform for iris",
    ui = function(id) {
      ns <- NS(id)
      tags$div(
        numericInput(ns("n_rows"), "Subset n rows", value = 6, min = 1, max = 150, step = 1)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
                 {
                   iris <- head(iris, num_rows)
                 },
                 num_rows = input$n_rows
          )
        })
      })
    }
  )
)

app1 <- init(
  data = teal_data(iris = iris),
  modules = example_module(datanames = c("iris", "mtcars"))
)

shinyApp(app1$ui, app1$server)

app2 <- init(
  data = teal_data(iris = iris),
  modules = example_module(datanames = c("iris", "mtcars"), transformers = my_transformers)
)
shinyApp(app2$ui, app2$server)
```

</details>

# No transformers - warning

no_transformers

# Transformers exist - no warning

transformers

# Results of local tests

<details><summary>Results of local tests</summary>

```r
> devtools::test()
ℹ Testing teal
[INFO] 2024-08-14 11:40:00.4037 pid:11652 token:[] teal You are using teal version 0.15.2.9052
✔ | F W  S  OK | Context
✔ |         10 | init [1.1s]                                                           
✔ |      6 129 | module_teal [57.3s]                                                   
✔ |         95 | modules                                                               
✔ |          7 | rcode_utils                                                           
✔ |          8 | report_previewer_module                                               
✔ |      6   0 | shinytest2-data_summary                                               
✔ |      3   0 | shinytest2-filter_panel                                               
✔ |      3   0 | shinytest2-init                                                       
✔ |      5   0 | shinytest2-landing_popup                                              
✔ |      4   0 | shinytest2-module_bookmark_manager                                    
✔ |      4   0 | shinytest2-modules                                                    
✔ |      5   0 | shinytest2-reporter                                                   
✔ |      1   0 | shinytest2-show-rcode                                                 
✔ |      6   0 | shinytest2-teal_data_module                                           
✔ |      2   0 | shinytest2-teal_slices                                                
✔ |      1   0 | shinytest2-utils                                                      
✔ |      2   0 | shinytest2-wunder_bar                                                 
✔ |         16 | teal_data_module-eval_code                                            
✔ |          4 | teal_data_module                                                      
✔ |         25 | teal_reporter                                                         
✔ |         15 | teal_slices-store                                                     
✔ |         18 | teal_slices                                                           
✔ |         36 | utils [9.7s]                                                          
✔ |         17 | validate_has_data                                                     
✔ |         36 | validate_inputs                                                       

══ Results ════════════════════════════════════════════════════════════════════════════
Duration: 74.8 s

── Skipped tests (48) ─────────────────────────────────────────────────────────────────
• need a fix in a .slicesGlobal (1): test-module_teal.R:1177:11
• testing depth 3 is below current testing specification 5 (42):
  test-shinytest2-data_summary.R:2:3, test-shinytest2-data_summary.R:21:3,
  test-shinytest2-data_summary.R:40:3, test-shinytest2-data_summary.R:62:3,
  test-shinytest2-data_summary.R:92:5, test-shinytest2-data_summary.R:139:3,
  test-shinytest2-filter_panel.R:5:3, test-shinytest2-filter_panel.R:32:3,
  test-shinytest2-filter_panel.R:71:3, test-shinytest2-init.R:5:3,
  test-shinytest2-init.R:15:3, test-shinytest2-init.R:70:3,
  test-shinytest2-landing_popup.R:5:3, test-shinytest2-landing_popup.R:25:3,
  test-shinytest2-landing_popup.R:43:3, test-shinytest2-landing_popup.R:67:5,
  test-shinytest2-landing_popup.R:146:3,
  test-shinytest2-module_bookmark_manager.R:5:3,
  test-shinytest2-module_bookmark_manager.R:19:3,
  test-shinytest2-module_bookmark_manager.R:33:3,
  test-shinytest2-module_bookmark_manager.R:44:3, test-shinytest2-modules.R:5:3,
  test-shinytest2-modules.R:44:3, test-shinytest2-modules.R:60:3,
  test-shinytest2-modules.R:77:3, test-shinytest2-reporter.R:5:3,
  test-shinytest2-reporter.R:25:3, test-shinytest2-reporter.R:48:3,
  test-shinytest2-reporter.R:87:3, test-shinytest2-reporter.R:103:3,
  test-shinytest2-show-rcode.R:5:3, test-shinytest2-teal_data_module.R:5:3,
  test-shinytest2-teal_data_module.R:42:3, test-shinytest2-teal_data_module.R:78:3,
  test-shinytest2-teal_data_module.R:130:3, test-shinytest2-teal_data_module.R:175:3,
  test-shinytest2-teal_data_module.R:215:3, test-shinytest2-teal_slices.R:5:3,
  test-shinytest2-teal_slices.R:48:3, test-shinytest2-utils.R:5:3,
  test-shinytest2-wunder_bar.R:5:3, test-shinytest2-wunder_bar.R:24:3
• todo (5): test-module_teal.R:1442:7, test-module_teal.R:1449:5,
  test-module_teal.R:1452:5, test-module_teal.R:1710:5, test-module_teal.R:1768:5

[ FAIL 0 | WARN 0 | SKIP 48 | PASS 416 ]
```

</details>

### CODE REVIEW

File: R/init.R [hunk 1]
```diff
@@ -224,7 +224,7 @@ init <- function(data,
     }
 
     is_modules_ok <- check_modules_datanames(modules, .teal_data_datanames(data))
-    if (!isTRUE(is_modules_ok)) {
+    if (!isTRUE(is_modules_ok) && length(unlist(extract_transformers(modules))) == 0) {
       lapply(is_modules_ok$string, logger::log_warn)
```
Comment (gogonzo) [inline R/init.R:228]: If you change it to warning you'd be able to `testthat::expect_warning`
```suggestion
lapply(is_modules_ok$string, warning, call. = FALSE)
```

**Review by gogonzo — COMMENTED** [review]
Please add a test

File: R/teal_data_module.R [hunk 2]
```diff
@@ -129,3 +129,17 @@ teal_transform_module <- function(ui, server, label = "transform module") {
     class = c("teal_transform_module", "teal_data_module")
   )
 }
+
+
+#' Extract all `transformers` from `modules`.
+#'
+#' @param modules `teal_modules` or `teal_module`
+#' @return A list of `transformers` nested the same way as input `modules`.
```
Comment (gogonzo) [inline R/teal_data_module.R:137]:
```suggestion
#' @return A list of `teal_transform_module` nested in the same way as input `modules`.
```

**Comment by m7pr** [comment]
> Please add a test

On it!

**Comment by m7pr** [comment]
Hey @gogonzo I added a test in `init` for the case with incomplete `datanames` but with `transformers` passed in `modules`. I also edited the text for the existing test that verified that in the previous scenario. 

I tried to check if we have any more tests verifying this warning message displayed on UI, but couldn't find any. Maybe we could extend tests with one or two `shinytest2` tests?

**Comment by m7pr** [comment]
I also see a test in `test-module_teal.R` that uses `teal_modules-module_1-validate_datanames-shiny_warnings`. The test starts with `testthat::it("throws warning when dataname is not available"`. I think we could add a test in there as well.

File: tests/testthat/test-init.R [hunk 3]
```diff
@@ -56,17 +56,48 @@ testthat::test_that("init throws when an empty `data` is used", {
   )
 })
 
-testthat::test_that("init throws warning when datanames in modules incompatible w/ datanames in data", {
-  testthat::local_mocked_bindings(log_warn = warning, .package = "logger")
+testthat::test_that(
+  "init throws warning when datanames in modules incompatible w/ datanames in data and there is no transformers",
+  {
+    testthat::local_mocked_bindings(log_warn = warning, .package = "logger")
```
Comment (gogonzo) [inline tests/testthat/test-init.R:62]: Is it needed? I thought `tests/testthat/setup-logger.R` turns off logger in tests

File: tests/testthat/test-init.R [hunk 4]
```diff
@@ -56,17 +56,48 @@ testthat::test_that("init throws when an empty `data` is used", {
...
+      ),
+      "Dataset \"iris\" is missing for tab 'example teal module'. Dataset available in data: \"mtcars\"."
+    )
+  }
+)
+
+testthat::test_that(
+  "init does not throw warning when datanames in modules incompatible w/ datanames in data and there are transformers",
+  {
+    testthat::local_mocked_bindings(log_warn = warning, .package = "logger")
```
Comment (gogonzo) [inline tests/testthat/test-init.R:77]: same here about warning vs logger

**Comment by gogonzo** [comment]
> I also see a test in `test-module_teal.R` that uses `teal_modules-module_1-validate_datanames-shiny_warnings`. The test starts with `testthat::it("throws warning when dataname is not available"`. I think we could add a test in there as well.

I don't know if it is feasible. This PR is about an `init` where `init` throws a warning and passes a data further to `srv_teal`. In `test-module_teal.R` you'll have to skip init and just pass insufficient data, which is already tested.

File: tests/testthat/test-init.R (hunk 3 above)
Comment (m7pr) [reply tests/testthat/test-init.R]: Yeah, looks like some old remaining. This works without mocking. I though it is needed to see the actual warning message. Just removed those https://github.com/insightsengineering/teal/pull/1319

**Comment by m7pr** [comment]
Yeah, but we also modified `module_teal_data` https://github.com/insightsengineering/teal/pull/1319/files#diff-8f20f15c7749d517fd977941dfda6eaa293268f76f57eb733bf4971273ea4e68R165 so the `shiny_warnings` is restricted with the extended condition. That's why I though an extra test https://github.com/insightsengineering/teal/pull/1319/commits/d2a658fa74f2286ec813bd7d1936987aa1cc4832 in there could be helpful

**Comment by gogonzo** [comment]
> Yeah, but we also modified `module_teal_data` https://github.com/insightsengineering/teal/pull/1319/files#diff-8f20f15c7749d517fd977941dfda6eaa293268f76f57eb733bf4971273ea4e68R165 so the `shiny_warnings` is restricted with the extended condition. That's why I though an extra test [d2a658f](https://github.com/insightsengineering/teal/commit/d2a658fa74f2286ec813bd7d1936987aa1cc4832) in there could be helpful

Good point, could you please include test in `test-teal_module.R`

**Comment by m7pr** [comment]
Yes @gogonzo - it's already included https://github.com/insightsengineering/teal/commit/d2a658fa74f2286ec813bd7d1936987aa1cc4832 in `test-module_teal.R`

**Review by gogonzo — CHANGES_REQUESTED** [review]
Error is not shown when data returned from `teal_transform_module` doesn't contain sufficient data:

<details>
<summary>app</summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

# pkgload::load_all("teal.data")
pkgload::load_all("teal")

anl_transformers <- list(
  teal_transform_module(
    label = "ANL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL (ADSL + ADTTE)"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    }
  ),
  teal_transform_module(
    label = "ANL",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("This transformer adds ANL2 (ADSL + ADTTE)"),
        numericInput(ns("obs"), "Number of subjects", value = 400, min = 0, max = 400)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
            {
              ANL2 <- dplyr::inner_join(
                head(ADSL, nobs),
                ADTTE[c("STUDYID", "USUBJID", setdiff(colnames(ADTTE), colnames(ADSL)))],
                by = c("USUBJID", "STUDYID")
              )
            },
            nobs = input$obs
          )
        })
      })
    }
  )
)

data <- teal_data_module(
  once = FALSE,
  ui = function(id) {
    ns <- NS(id)
    tagList(
      numericInput(ns("obs"), "Number of observations to show", 1000),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      logger::log_trace("example_module_transform2 initializing.")
      eventReactive(input$submit, {
        data <- teal_data(a = NULL) |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- head(teal.data::rADSL, n = n)
              ADTTE <- teal.data::rADTTE
              iris <- iris
              iris_raw <- iris
              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            n = as.numeric(input$obs)
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("CO2", "ADTTE", "iris_raw", "iris", "ADSL")
        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = list(
    example_module("insufficient+transforms", datanames = c("ADTTE", "insufficient"), transformers = anl_transformers),
    reporter_previewer_module()
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    include_varnames = list(
      ADSL = c("SEX", "AGE")
    )
  )
)

runApp(app)
``` 
</details>

**Comment by m7pr** [comment]
@gogonzo maybe there is a missunderstanding in what I thought we wanted to achieve with this issue.
If any transformers are passed, then insufficient data warning will not be shown.

I treat the conditions from the initial card as `AND` condition
image

So the warning will be displayed if both are satisified. Here you have transformers in `example_module` so warning is omitted.

Sorry if I got this wrong. Happy to discuss when I'm back from holidays.

**Comment by gogonzo** [comment]
> Here you have transformers in example_module so warning is omitted.

We shouldn't omit a warning if data returned to the module didn't produce sufficient datanames. Issue is to skip warning **on init** if transform is provided and not to ignore datanames warning completely when a module has any transformer.

**Comment by m7pr** [comment]
Make sense @gogonzo and thank you for the clarification. Will be able to edit the PR today

**Comment by m7pr** [comment]
Sorry for the initial extension of what we planned. Initially I provided restriction to init and to the UI of the module. I just removed the restriciton from the UI. The warning is not displayed in the init, but is visible in the UI.

<details><summary>Results of tests</summary>

```r
[ FAIL 0 | WARN 0 | SKIP 48 | PASS 417 ]
```

</details>

<details><summary>Tested with</summary>

This app https://github.com/insightsengineering/teal/pull/1319#pullrequestreview-2239860749
and the app below

```r
pkgload::load_all('teal.data')
pkgload::load_all('teal.slice')
pkgload::load_all('teal')

my_transformers <- list(
  teal_transform_module(
    label = "Custom transform for iris",
    ui = function(id) {
      ns <- NS(id)
      tags$div(
        numericInput(ns("n_rows"), "Subset n rows", value = 6, min = 1, max = 150, step = 1)
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          within(data(),
                 {
                   iris <- head(iris, num_rows)
                 },
                 num_rows = input$n_rows
          )
        })
      })
    }
  )
)

app1 <- init(
  data = teal_data(iris = iris),
  modules = example_module(datanames = c("iris", "mtcars"))
)

shinyApp(app1$ui, app1$server)

app2 <- init(
  data = teal_data(iris = iris),
  modules = example_module(datanames = c("iris", "mtcars"), transformers = my_transformers)
)
shinyApp(app2$ui, app2$server)
```

</details>

<details><summary>Print screens of functionality in action</summary>

app1 - no transformers - warning displayed
app2 - transformers exist - warning not displayed
2024_08_27_CONSOLE

app2 - transformers exist - warning displayed in UI
2024_08_27_UI

Second app - no warning on init, visible warning on UI

image


</details>

**Comment by m7pr** [comment]
Thanks for the patience @gogonzo - just checked with 2 apps and run tests locally. This now only displays the warning in UI and hides the warning for the console when you run `init`.

**Review by gogonzo — APPROVED** [review]
👍

**Pull request merged**